File: src/main/java/com/sgx/icms/service/ManagerService.java
package com.sgx.icms.service;

import java.math.BigDecimal;

import com.sgx.icms.dao.ClaimDao;
import com.sgx.icms.dao.SettlementDao;
import com.sgx.icms.db.Db;
import com.sgx.icms.domain.AuditLog;
import com.sgx.icms.domain.Settlement;
import com.sgx.icms.web.support.SessionUser;

/** Manager-specific dashboard stats and settlement override. */
public class ManagerService {

    private final SettlementDao settlementDao = new SettlementDao();
    private final ClaimDao claimDao = new ClaimDao();
    private final AuditService audit = new AuditService();

    /** [pendingApproval, highRisk, slaBreaches, settled] for the manager dashboard. */
    public long[] dashboardStats() {
        return Db.withConnection(conn -> new long[] {
            Db.queryLong(conn, "SELECT COUNT(*) FROM claims WHERE status = 'PENDING_APPROVAL'"),
            Db.queryLong(conn, "SELECT COUNT(*) FROM claims WHERE risk_level = 'HIGH' "
                    + "AND status NOT IN ('SETTLED','CLOSED','REJECTED','WITHDRAWN')"),
            Db.queryLong(conn, "SELECT COUNT(*) FROM claims WHERE sla_due_date < CURDATE() "
                    + "AND status NOT IN ('SETTLED','CLOSED','REJECTED','WITHDRAWN')"),
            Db.queryLong(conn, "SELECT COUNT(*) FROM claims WHERE status IN ('SETTLED','CLOSED')")
        });
    }

    /**
     * Manager overrides the settlement amount (must already exist, created by the agent
     * post-approval).
     * @throws IllegalStateException if there is no settlement to override
     */
    public void overrideSettlement(SessionUser manager, long claimId, BigDecimal amount,
                                   String justification, String ip) {
        if (amount == null || amount.signum() <= 0) {
            throw new IllegalArgumentException("Enter a valid override amount.");
        }
        Db.inTransaction(conn -> {
            Settlement s = settlementDao.findByClaim(conn, claimId);
            if (s == null) {
                throw new IllegalStateException("No settlement exists yet to override.");
            }
            settlementDao.updateAmount(conn, claimId, amount, justification);
            com.sgx.icms.domain.Claim claim = claimDao.findById(conn, claimId);
            AuditLog a = new AuditLog();
            a.setUserId(manager.getId());
            a.setUsername(manager.getUsername());
            a.setRole(manager.getRole());
            a.setAction("SETTLEMENT_OVERRIDE");
            a.setEntity((claim == null ? ("claim:" + claimId) : claim.getClaimNo()) + " (₹" + amount + ")");
            a.setIpAddress(ip);
            a.setResult(AuditLog.RESULT_SUCCESS);
            audit.record(conn, a);
            return null;
        });
    }
}

File: backend/src/services/managerService.js
'use strict';

const { knex } = require('../db/knex');
const { withTransaction } = require('../db/tx');
const settlementRepo = require('../repositories/settlementRepo');
const claimRepo = require('../repositories/claimRepo');
const audit = require('./auditService');
const { toAmount } = require('../utils/money');
const { ValidationError, ConflictError } = require('../utils/errors');

/** Manager dashboard stats + settlement override — Node port of ManagerService. */

async function dashboardStats() {
  const q = async (sql) => {
    const [rows] = await knex.raw(sql);
    return Number(Object.values(rows[0])[0]);
  };
  const [pendingApproval, highRisk, slaBreaches, settled] = await Promise.all([
    q("SELECT COUNT(*) FROM claims WHERE status = 'PENDING_APPROVAL'"),
    q("SELECT COUNT(*) FROM claims WHERE risk_level = 'HIGH' AND status NOT IN ('SETTLED','CLOSED','REJECTED','WITHDRAWN')"),
    q("SELECT COUNT(*) FROM claims WHERE sla_due_date < CURDATE() AND status NOT IN ('SETTLED','CLOSED','REJECTED','WITHDRAWN')"),
    q("SELECT COUNT(*) FROM claims WHERE status IN ('SETTLED','CLOSED')"),
  ]);
  return { pendingApproval, highRisk, slaBreaches, settled };
}

async function overrideSettlement(manager, claimId, amount, justification, ip) {
  const amountNum = Number(amount);
  if (amount == null || amount === '' || !Number.isFinite(amountNum) || amountNum <= 0) {
    throw new ValidationError('Enter a valid override amount.');
  }
  const amt = toAmount(amount);
  return withTransaction(async (trx) => {
    const s = await settlementRepo.findByClaim(trx, claimId);
    if (!s) throw new ConflictError('No settlement exists yet to override.');
    await settlementRepo.updateAmount(trx, claimId, amt, justification);
    const claim = await claimRepo.findById(trx, claimId);
    await audit.record(trx, {
      userId: Number(manager.id),
      username: manager.username,
      role: manager.role,
      action: 'SETTLEMENT_OVERRIDE',
      entity: `${claim ? claim.claimNo : 'claim:' + claimId} (₹${amt})`,
      ipAddress: ip,
      result: 'SUCCESS',
    });
  });
}

module.exports = { dashboardStats, overrideSettlement };
